Create generating-tls-files-yourself-to-avoid-providing-a-fingerprint.md (#103)
exasol/public-knowledgebase: Environment-Management/generating-tls-files-yourself-to-avoid-providing-a-fingerprint.md

* [Upload TLS Certificate, version 7.1](https://docs.exasol.com/db/7.1/administration/on-premise/access_management/tls_certificate.htm)

The description of the TLS encryption feature could be found in the following changelog entry: [CHANGELOG: TLS for all Exasol drivers](https://exasol.my.site.com/s/article/Changelog-content-6507?language=en_US).

Our github repository "Exasol Java tutorial" also contains two articles explaining the overall concept and relation to Exasol:

* [An Introduction to TLS](https://github.com/exasol/exasol-java-tutorial/blob/main/tls-tutorial/doc/tls_introduction.md)
* [TLS With Exasol](https://github.com/exasol/exasol-java-tutorial/blob/main/tls-tutorial/doc/tls_with_exasol.md)

Please acquaint yourself with the process of using and uploading TLS certificates for your database version, as well as with the description of the TLS encryption feature, using the links above.

The tutorial applies to a particular use case:

1. Only a connection string with IP addresses or DNS names could be used, fingerprint couldn't be added.

> 2. Customer's IT Department or Security Department couldn't purchase a certificate from a public certification authority (CA) and also couldn't generate certificate in-house and sign it with company's customr root CA certificate.

Author: ```suggestion
2. Customer's IT Department or Security Department couldn't purchase a certificate from a public certification authority (CA) and also couldn't generate certificate in-house and sign it with company's customer root CA certificate.
```

## How to generate TLS files

All settings here could be hardened by customer if needed by customer's security policies.

File names could also be changed when needed, but in a coordinated way as some commands depend on files generated on the previous steps.

We'll be listing commands and describe the output.

Full explanation of `openssl` utility parameters could be found in `openssl` documentation: [OpenSSL commands](https://docs.openssl.org/master/man1/).

```shell
# Generate CA key
openssl genrsa -out rootCA_key.pem 4096
```

This command creates file `rootCA_key.pem`. Its content is sensitive and should be accessible by a limited subset of administrative users.

```shell
# Create CA cert
openssl req -x509 -new -nodes -key rootCA_key.pem -sha256 -days 3650 -out rootCA_cert.pem -subj '/CN=Custom_root_CA_for_Exasol_databases/O=test/C=DE'
```

This command creates a CA certificate as file `rootCA_cert.pem` with the following settings

* Valid for 10 years (`-days 3650`).
* The key is not password protected (`-nodes`)
* Common name (`CN` in `-subj`) is "Custom root CA for Exasol databases"
* Organization Name (`O` in `-subj`) is "test"

> * Country Name (`C` in `-subj`) is "test"

Author: ```suggestion
* Country Name (`C` in `-subj`) is "DE"
```

```shell
# Create server cert
openssl x509 -req -in server_cert.csr -CA rootCA_cert.pem -CAkey rootCA_key.pem -CAcreateserial -out server_cert.pem -days 3650 -sha256 -extfile server_cert.ext
```

This command finally creates server certificate file `server_cert.pem`. This file is not sensitive.

We are done with certificate generation. Next one needs to create a certificate chain file by concatenating files `server_cert.pem` and `rootCA_cert.pem`

```shell
cat server_cert.pem > cert_chain.pem; cat rootCA_cert.pem >> cert_chain.pem
```

and uploading the resulting certificate chain file `cert_chain.pem` and key file `server_key.pem` via respective database interfaces.

## Configuration of selected client applications

In this section we'll show how to configure some client tools to work with the certificates we've just created.
One needs to do so as the generated root CA certificate is not known to client applications.

> ### DBeaver, DB Visualizer and other third party tools using Exasol JDBC driver

Author: ```suggestion
### DBeaver, DbVisualizer and other third party tools using Exasol JDBC driver
```

Third party SQL clients leveraging Exasol JDBC driver typically come with their own JRE (Java Runtime Environment).
By design Java applications use their own truststore and check if they trust database certificates by searching it in such a truststore.

> For example, as of 2024-11-25 for DBeaver (steps and paths for DB Visualizer are similar) JRE resides in `jre` subfolder of installation folder.

Author: ```suggestion
For example, as of 2024-11-25 for DBeaver (steps and paths for DbVisualizer are similar) JRE resides in `jre` subfolder of installation folder.
```
